### DISCUSSION TITLE
Questions About Current State, My Needs, and Whether it Makes Sense for Me to Contribute

### DISCUSSION THREAD

**Discussion by zicklag** [subconsciousnetwork/noosphere#605] [Q&A]
Hey there!

I'm just starting to look into Noosphere and I had some questions. I have a collaborative project with some friends that I've decided I need a collaborative Zettelkasten / linked index-card note taking system for. I'm going to develop the user experience myself, because I don't know of anything that already exists that will do what I want, and I'm investigating the technology to build the database/server of sorts that it will use.

My needs are:

- I can have a personal, private set of notes that I can keep local and backup to whatever storage I desire.
- I can have an authenticated, collaborative note cluster that me and my friends can all edit in realtime. 
- I can link from my personal notes to my collaborative notes, and view them all in one place, even though my friends can't see my personal notes.
- I can use this on a network that blocks access to 99% of the internet including the IPFS network.
- Related to the network access restriction, I have to be able to host a storage / coordination node on my own server that I _can_ get to so that me and my friends can store our collaborative notes on it.

Noosphere is very close in vision to what I need, but I'm not sure what state it's currently in or whether  I can just grab some of your Rust crates and use them to start hosting a private Noosphere.

I'm looking to get something stood up as soon as I can, so my other consideration was just throwing together something with [`yrs`](https://docs.rs/yrs/latest/yrs/) as a CRDT for syncing and use a key-value store for storage, maybe roll my own simple content addressing and storage scheme, knowing that with simple subtext note content I can always migrate my storage strategy.

Because I'm wanting to really limit the scope to simple notes and a graph visualization tool, I'm not really scared of the scope for rolling my own solution, but I'm wondering if it might be worth trying to use noosphere and contributing if necessary.

My initial impression is that noosphere isn't in a state yet where it will be super easy to plugin-in and since I've got a different use case than the primary noosphere vision, what I'm needing might not line up perfectly with what development needs to be done right now.

So I'm leaning towards just rolling my own, even if the backend is temporary, but I wanted to open the discussion to see what your thoughts were on it, and whether you think noosphere might be a good fit.

**Comment by justinabrahms** [comment]
This is my understanding:

>    I can have a personal, private set of notes that I can keep local and backup to whatever storage I desire.

This is something we're after too.

>    I can have an authenticated, collaborative note cluster that me and my friends can all edit in realtime.

Realtime, multi-writer editing of the same note (e.g. google docs) isn't on the roadmap right now. Shared spheres wherein multiple writers can simultaneously edit different notes is expected.

>    I can link from my personal notes to my collaborative notes, and view them all in one place, even though my friends can't see my personal notes.
This is expected.

>    I can use this on a network that blocks access to 99% of the internet including the IPFS network.
Clients connect to a gateway. The gateway speaks to IPFS today. Your client doesn't need access to IPFS, but I believe that the gateway expects IPFS. I don't know how likely that is to change.

>    Related to the network access restriction, I have to be able to host a storage / coordination node on my own server that I can get to so that me and my friends can store our collaborative notes on it.

This is supported if you want you and your friends to be an island. If you want this to also tie into the main network, it's on the books but not there yet.

cc @cdata who can confirm my understanding of the above.

**Reply by zicklag** [reply]
Thanks for the reply!

For realtime editing of the same note, I might be able to accomplish that by making the realtime server a "user" that has access to the shared sphere, so when users want to collaboratively edit, they talk to the realtime server to create the room for the note, and then the server submits the final update to the note in the sphere.

Now that I say that, though, if the realtime server is just a user with it's own DID, then we don't even need a shared sphere in that case. The realtime server would handle authentication and writing to the notes.

> This is supported if you want you and your friends to be an island.

For now that's fine. It'd be cool to integrate with the broader noosphere at some point, but right now it's fine if it's local to us.

It's sounding like maybe we're not too far off...

---

I did some more looking into the rust crate API docs, and I'm liking what I'm seeing. One concern of mine was being able to configure networking, storage, and the IPFS implementation myself, but it looks like you've got the necessary abstractions like the [`Storage`](https://docs.rs/noosphere-storage/0.8.1/noosphere_storage/trait.Storage.html) trait. I haven't figured out how to do custom networking yet, but it looks like it's in there, if that's not part of `Storage` too.

As far as I can tell you aren't locked in to IPFS either, so I'm free to use something like [Iroh](https://github.com/n0-computer/iroh) embedded in the client, instead of accessing an IPFS gateway over HTTP.

I'm really liking a lot of the ideas I'm seeing here, like UCANs, etc. I think I'll be able to use noosphere for my app, even if it's a little unconventionally. If I need to or some things I figure out end up being stuff that makes sense to contribute to noosphere, then I might be able to open discussions or PRs for that stuff, too.

I'm excited to check this out!
